src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -0,0 +1,257 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.CustomProtocol;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using CodeAction = Microsoft.CodeAnalysis.CodeActions.CodeAction;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    /// <summary>
+    /// Resolves a code action by filling out its Edit or Command property.
+    /// </summary>
+    [ExportLspMethod(MSLSPMethods.TextDocumentCodeActionResolveName), Shared]
+    internal class CodeActionResolveHandler : AbstractRequestHandler<LSP.VSCodeAction, LSP.VSCodeAction>
+    {
+        private readonly ICodeFixService _codeFixService;
+        private readonly ICodeRefactoringService _codeRefactoringService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CodeActionResolveHandler(
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            ILspSolutionProvider solutionProvider)
+            : base(solutionProvider)
+        {
+            _codeFixService = codeFixService;
+            _codeRefactoringService = codeRefactoringService;
+        }
+
+        public override async Task<LSP.VSCodeAction> HandleRequestAsync(
+            LSP.VSCodeAction codeAction,
+            LSP.ClientCapabilities clientCapabilities,
+            string? clientName,
+            CancellationToken cancellationToken)
+        {
+            var data = ((JToken)codeAction.Data).ToObject<CodeActionResolveData>();
+            var document = SolutionProvider.GetDocument(data.TextDocument, clientName);
+            var codeActions = await CodeActionsHandler.GetCodeActionsAsync(
+                document,
+                _codeFixService,
+                _codeRefactoringService,
+                data.Range,
+                cancellationToken).ConfigureAwait(false);
+
+            if (codeActions == null || !codeActions.Any())
+            {
+                return codeAction;
+            }
+
+            var codeActionToResolve = GetCodeActionToResolve(data.DistinctTitle, codeActions);
```

Comment:
so to resolve a code action, we have to compute all the code actions again?  seems costly.  can/should we store/save things locally across calls to speed this up?

Reply:
One suggestion I would have is to do the caching in a follow-up pr.  This matches the current behavior in nexus (always recompute) so we're not regressing anything.

This pr is already fairly big, and caching I think will be a little complex and would prefer to review separately.

Reply:
i am ok with that decision for this PR. :)

Reply:
Opened #45727 to track this.